lib/queue.js

```diff
@@ -830,16 +830,40 @@ class Queue extends Emitter {
       args[i] = arguments[i];
     }
 
-    return this._commandable().then((client) => {
-      // Get the sha for the script after we're ready to avoid a race condition
-      // with the lua script loader.
-      args[0] = lua.shas[name];
-
-      const promise = helpers.deferred();
-      args.push(promise.defer());
-      client.evalsha.apply(client, args);
-      return promise;
-    });
+    return this._commandable()
+      .then((client) => {
+        // Get the sha for the script after we're ready to avoid a race
+        // condition with the lua script loader.
+        args[0] = lua.shas[name];
+
+        const promise = helpers.deferred();
+        args.push(promise.defer());
+        client.evalsha.apply(client, args);
+        return promise;
+      })
+      .catch((err) => {
+        // In cases where the Redis server fails over, the script cache is
+        // cleared (even when the cache itself is still intact.) We detect this
+        // condition and reload the scripts.
+        if (
+          err.name === 'ReplyError' &&
+          err.message === 'NOSCRIPT No matching script. Please use EVAL.'
+        ) {
+          return lua.buildCache(this.client).then((client) => {
```

Comment:
I'm somewhat concerned that this will hammer the redis server if the client tries to execute a bunch of scripts all at once - they'll all come back with `NOSCRIPT` errors and will all separately trigger this code path. IMO we should handle that by introducing new code (either in `buildCache` or between here and `buildCache`) to memoize the promise. It'd want to be per-client, I suspect - maybe using a `WeakMap` (just storing it on the `Queue` would work too)?

Reply:
[I believe `WeakMap` requires node versions >= 12 (10 behind an interpreter flag)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap#Browser_compatibility); does this change this recommendation at all? I see that you are planning deprecation of node 4 & 6 - wondering about 8/10 in this context & given that your friends at Mixmax (probably) aren't the only users of this package. I know I personally use 10 in some places still :L (tho this could be the kick in the butt to prioritize some other things.)

Reply:
I'm seeing Node `0.12`, not `12` - am I looking at something different?

Reply:
nah, you're correct - I mean 0.12 and not 12. d'oh, I'm just mixed up.